src/psyclone/psyGen.py

```diff
@@ -3083,6 +3083,129 @@ def scalars(self):
+        '''Determine whether the accesses to the provided field overlap
+        with the accesses of the source argument
+
+        We do not currently deal with accesses to a subset of a field
+        (unless it is a vector). This distinction will need to be
+        added once loop splitting is supported and would be useful for
+        boundary condition kernels
+
+        :param arg: the argument to compare with our internal argument
+        :type arg: :py:class:`psyclone.psyGen.Argument`
+        :return Bool: True is there are overlapping accesses between \
+                      arguments and False if not.
+
+        '''
+        if self._arg.name != arg.name:
+            # the arguments are different args so do not overlap
+            return False
+
+        if isinstance(self._call, HaloExchange) and \
+           isinstance(arg.call, HaloExchange) and \
+           (self._arg.vector_size > 1 or arg.vector_size > 1):
+            # This is a vector field and both accesses come from halo
+            # exchanges. As halo exchanges only access a particular
+            # vector the accesses do not overlap if the vector indices
+            # being accessed differ.
+
+            # sanity check
+            self._call.check_vector_halos_differ(arg.call)
+
+            if self._call.vector_index != arg.call.vector_index:
+                # accesses are to different vector indices so do not overlap
+                return False
+        # accesses do overlap
+        return True
+
+    def reset_coverage(self):
+        '''Reset internal state to allow re-use of the object for a different
+        situation.
+
+        '''
+        self._covered = False
+        self._vector_index_access = []
+
+    def update_coverage(self, arg):
+        '''Record any overlap between accesses to the supplied argument and
+        the source argument. If the overlap results in all of the
+        accesses to the source argument being covered (either directly
+        or as a combination with previous arguments) then ensure that
+        the covered() method returns True.
+
+        :param arg: the argument to compare with our internal argument \
+                    and update coverage information
+        :type arg: :py:class:`psyclone.psyGen.Argument`
+
+        '''
+
+        if not self.overlaps(arg):
+            ''' there is no overlap so there is nothing to update '''
+            return
+
+        if isinstance(arg.call, HaloExchange) and \
+           self._arg.vector_size > 1:
+            # The supplied argument is a vector field coming from a
+            # halo exchange and therefore only accesses one of the
+            # vectors
+
+            if isinstance(self._call, HaloExchange):
+                # I am also a halo exchange so only access one of the
+                # vectors. At this point the vector indices of the two
+                # halo exchange fields must be the same, which should
+                # never happen.
```

Comment:
because of the check on `self.overlaps(arg)` earlier?

Reply:
Yes.